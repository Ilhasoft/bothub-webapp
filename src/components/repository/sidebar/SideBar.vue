<template>
  <div class="sidebar">
    <div :class="colapse ? 'sidebar-wrapper' : 'sidebar-wrapper--colapsed'">
      <div
        v-if="colapse"
        ref="colapseButton"
        class="sidebar-wrapper__colapse-button"
        @click="colapseHandle()">
        <b-icon icon="arrow-right"/>
      </div>
      <b-menu v-if="colapse">
        <div class="menu-header">
          <img
            src="~@/assets/imgs/logo.svg"
            alt="bothub">
        </div>
        <b-menu-list>
          <template
            slot="label"
            slot-scope="props">
            Test
            <b-icon
              :icon="props.expanded ? 'caret-down' : 'caret-up'"
              class="is-pulled-right"/>
          </template>
          <b-menu-item
            :to="{ name: 'repository-summary' }"
            tag="router-link"
            icon="home"
            size="is-medium"
            label="Summary"/>
          <b-menu-item
            :to="{ name: 'repository-training' }"
            tag="router-link"
            icon="refresh"
            label="Training"/>
          <b-menu-item
            :active="isTestsActive"
            :expanded="isTestsActive"
            icon="wechat"
            @click="isTestsActive = !isTestsActive">
            <template
              slot="label"
              slot-scope="props">
              <span>Test</span>
              <b-icon
                :icon="props.expanded ? 'caret-down' : 'caret-up'"
                class="is-pulled-right"/>
            </template>
            <b-menu-item
              :to="{ name: 'repository-test' }"
              tag="router-link"
              label="Sentences"/>
            <b-menu-item
              :to="{ name: 'repository-test' }"
              tag="router-link"
              label="Results"/>
            <b-menu-item
              :to="{ name: 'repository-test' }"
              tag="router-link"
              label="Versions"/>
          </b-menu-item>
          <b-menu-item
            :to="{ name: 'repository-translate' }"
            tag="router-link"
            icon="translate"
            label="Translation"/>
          <b-menu-item
            :to="{ name: 'repository-integration' }"
            tag="router-link"
            icon="power-plug"
            label="Integration"/>
            <b-menu-item
            :active="isSettingsActive"
            :expanded="isSettingsActive"
            icon="settings"
            @click="isSettingsActive = !isSettingsActive">
            <template
              slot="label"
              slot-scope="props">
              <span>Settings</span>
              <b-icon
                :icon="props.expanded ? 'caret-down' : 'caret-up'"
                class="is-pulled-right"/>
            </template>
            <b-menu-item
              :to="{ name: 'repository-settings' }"
              tag="router-link"
              label="Settings"/>
            <b-menu-item
              :to="{ name: 'repository-versions' }"
              tag="router-link"
              label="Versions"/>
          </b-menu-item>
        </b-menu-list>
      </b-menu>
      <div
        v-else
        class="icon-list-wrapper">
        <div
          v-if="!colapse"
          class="colapse-button icon-list"
          @click="colapseHandle()">
          <b-icon
            icon="arrow-right-bold"/>
        </div>
        <b-icon
          class="icon-list"
          icon="home"
          @click.native="routerHandle('repository-summary')" />
        <b-icon
          class="icon-list"
          icon="wechat"
          @click.native="routerHandle('repository-training')" />
        <b-dropdown
          aria-role="list">
          <b-icon
            slot="trigger"
            class="icon-list"
            icon="chart-line" />
          <b-dropdown-item
            aria-role="listitem"
            @click="routerHandle('repository-test')">Action</b-dropdown-item>
          <b-dropdown-item
            aria-role="listitem"
            @click="routerHandle('repository-test')">Another action</b-dropdown-item>
          <b-dropdown-item
            aria-role="listitem"
            @click="routerHandle('repository-test')">Something else</b-dropdown-item>
        </b-dropdown>
        <b-icon
          class="icon-list"
          icon="translate"
          @click.native="routerHandle('repository-translate')" />
        <b-icon
          class="icon-list"
          icon="power-plug"
          @click.native="routerHandle('repository-integration')" />
        <b-icon
          class="icon-list"
          icon="settings"
          @click.native="routerHandle('repository-settings')" />
        
          
      </div>
      <div/>
    </div>
  </div>
</template>
<script>

export default {
  name: 'SideBar',
  data() {
    return {
      isSettingsActive: false,
      isTestsActive: false,
      colapse: true,
    };
  },
  methods: {
    routerHandle(path) {
      this.$router.push({
        name: `${path}`,
      });
    },
    colapseHandle() {
      this.$emit('colapse');
      this.colapse = !this.colapse;
    },
  },
};
</script>
<style lang="scss">
.icon-list-wrapper {
  text-align: center;
}

.menu-header {
  width: 100%;
  height: 5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;

  img {
    height: 1.5rem;
  }
}

.sidebar {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  z-index: 1;

  .sidebar-wrapper {
    background: #2F343D;
    height: 100%;
    padding: 1rem;
    width: 260px;
    transition: width .1s;

  &__colapse-button {
    background: #696968;;
    color:#FFFFFF;
    width: 2rem;
    height: 2rem;
    position: absolute;
    display: flex;
    align-items: center;
    top: 2rem;
    left: 0;
    cursor: pointer;

    span {
      margin-left: 3px;
    }
  }

    &--colapsed {
      transition: width .1s;
      background: #2F343D;
      height: 100%;
      padding: 1rem;
      width: 60px;
      color: #FFFFFF;
    }
  }
}

.icon-list {
  margin: .5rem 0;
  cursor: pointer;
}
.icon-list:hover {
  color: #12A391;
}
</style>
