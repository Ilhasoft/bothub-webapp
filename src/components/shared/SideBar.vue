<template>
  <div>
    <div class="sidenav__button-open">
      <bh-icon-button
        class="sidenav__button-open__icon"
        value="botinho"
        size="small"
        @click="openNav()"
      />
    </div>
    <div
      id="mySidenav"
      class="sidenav">
      <div class="sidenav__repository-info">
        <div class="sidenav__repository-info__big-badge">
          <bh-icon-button
            :value="repositoryInfoIcon"
            size="small"
            class="sidenav__repository-info__big-badge__icon" />
        </div>
        <div class="sidenav__repository-info__text-info">
          <strong>{{ repository.name }}</strong>
          <p>
            created by <strong class="text-color-primary">{{ repository.owner__nickname }}</strong>
          </p>
        </div>
      </div>
      <div>
        <button
          href="javascript:void(0)"
          class="sidenav__closebtn"
          @click="closeNav()">&times;</button>
        <router-link
          v-for="([name, label], i) in repositoryLinks"
          :key="i"
          :to="toFactory(name)"> {{ label }}
        </router-link>
      </div>
      <div class="sidenav__status">
        <a
          v-if="authenticated && repository.authorization.can_write && repository.ready_for_train"
          href="#">
          <span class="text-color-primary sidenav__status__info">
            <bh-icon
              value="botinho" />
            Your bot is ready to be trained </span>
        </a>
        <a
          v-else-if="!authenticated"
          href="#"
          @click.prevent="openLoginModal">
          Sign In
        </a>
        <a
          v-else-if="authenticated && repository.available_request_authorization"
          href="#"
          @click.prevent="openRequestAuthorizationModal">
          Request Authorization
        </a>
        <a
          v-else-if="authenticated
            && repository.request_authorization
          && !repository.request_authorization.approved_by">
          <p class="text-color-grey-dark">
            Authorization Requested
          </p>
        </a>
      </div>
      <div
        v-if="authenticated"
        class="sidenav__profile-menu">
        <a
          href="#"
          @click.prevent="openMyProfile">
          {{ myProfile.name || '...' }}
        </a>
        <a
          href="#"
          @click.prevent="openNewRepositoryModal">
          Start your bot
        </a>
        <a
          href="#"
          @click.prevent="logout">
          Logout
        </a>
      </div>
    </div>
    <new-repository-modal
      :active="newRepositoryModalOpen"
      @requestClose="openNewRepositoryModal()" />
    <request-authorization-modal
      v-if="repository"
      :open.sync="requestAuthorizationModalOpen"
      :repository-uuid="repository.uuid"
      @requestDispatched="onAuthorizationRequested()" />
  </div>
</template>
<script>
import RepositoryInfo from '@/components/repository/RepositoryInfo';
import NewRepositoryModal from '@/components-v1/shared/NewRepositoryModal';
import RequestAuthorizationModal from '@/components/repository/RequestAuthorizationModal';
import { mapGetters, mapActions } from 'vuex';


export default {
  name: 'VsSidebar',
  inject: {
    updateRepository: {
      default: null,
    },
  },
  components: {
    RepositoryInfo,
    NewRepositoryModal,
    RequestAuthorizationModal,
  },
  props: {
    repository: {
      type: Object,
      default: null,
    },
  },
  data() {
    return {
      repositoryLinks: [
        ['repository-summary', 'Summary'],
        ['repository-training', 'Training'],
        ['repository-test', 'Test'],
        ['repository-translate', 'Translate'],
        ['repository-translations-status', 'Translation status'],
        ['repository-integration', 'Integration'],
        ['repository-settings', 'Settings'],
      ],
      newRepositoryModalOpen: false,
      requestAuthorizationModalOpen: false,
    };
  },
  computed: {
    ...mapGetters([
      'myProfile',
      'authenticated',
    ]),

    repositoryInfoIcon() {
      return (this.repository.categories[0] && this.repository.categories[0].icon) || 'botinho';
    },
  },
  methods: {
    ...mapActions([
      'logout',
      'openLoginModal',
    ]),
    openNav() {
      document.getElementById('mySidenav').style.width = '280px';
    },
    closeNav() {
      document.getElementById('mySidenav').style.width = '0';
    },
    toFactory(name) {
      return {
        name,
        params: {
          ownerNickname: this.repository.owner__nickname,
          slug: this.repository.slug,
        },
      };
    },
    openMyProfile() {
      this.$router.push({ name: 'myProfile' });
    },
    openNewRepositoryModal() {
      this.newRepositoryModalOpen = !this.newRepositoryModalOpen;
    },
    openRequestAuthorizationModal() {
      this.requestAuthorizationModalOpen = true;
    },
    onAuthorizationRequested() {
      this.requestAuthorizationModalOpen = false;
      this.$toast.open({
        message: 'Request made! Wait for review of an admin.',
        type: 'is-success',
      });
      this.updateRepository(false);
    },
  },
};
</script>
<style lang="scss">
@import '~bh/src/assets/scss/colors.scss';


.sidenav {
  height: 100%;
  width: 0;
  position: fixed;
  z-index: 10;
  top: 0;
  left: 0;
  background-color: $color-fake-white;
  overflow-x: hidden;
  transition: 0.5s;

  &__button-open {
    margin: 1rem;
    $size: 3rem;
    position: relative;
    display: block;
    width: $size;
    height: $size;
    border-radius: 50%;
    background-color: #009688;
    overflow: hidden;

    &__icon {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      color: white;
    }
  }

  &__repository-info {
    display: grid;
    grid-template-columns: 5rem 1fr;
    padding: 0.5rem 0;

    &__big-badge {
      margin: 1rem;
      $size: 3rem;
      position: relative;
      display: block;
      width: $size;
      height: $size;
      border-radius: 50%;
      background-color: #009688;
      overflow: hidden;

      &__icon {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        color: white;
      }
    }

    &__text-info {
      text-align: left;
      align-self: center;
      padding-right: 2rem;
      p {
        margin: 0;
        font-size: 1rem;
      }
    }
  }

  a {
    padding: 1rem 1rem;
    text-decoration: none;
    border-bottom: 1px solid $color-grey;
    color: black;
    display: block;
    transition: 0.3s;
    text-align: center;

    &:last-child {
      border-bottom: 1px solid #009688;
    }

    &:hover {
      color: #009688;
    }
  }

  &__closebtn {
      border-bottom: 15px;
      position: absolute;
      top: 0;
      right: .05rem;
      font-size: 36px;
      margin-left: 50px;
      border:none;
      background-color: $color-fake-white;
    }

  &__profile-menu {
    a:last-child {
      border-bottom: 1px solid $color-grey;
    }
  }

  &__status {
    &__info {
      display: flex;
      align-items: center;
      justify-content: space-around;
    }
  }
}

@media screen and (max-height: 450px) {
  .sidenav {padding-top: 15px;}
  .sidenav a {font-size: 18px;}
}
</style>
