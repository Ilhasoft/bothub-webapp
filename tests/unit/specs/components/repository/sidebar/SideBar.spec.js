import { shallowMount, createLocalVue } from '@vue/test-utils';
import SideBar from '@/components/repository/sidebar/SideBar';
import Vuex from 'vuex';
import Buefy from 'buefy';

let store;

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Buefy);

const currentRepository = {
  name: 'bothub',
  owner__nickname: 'weni',
  uuid: '123456'
}

describe('SideBar.vue', () => {
  store = new Vuex.Store({
    modules: {
      Repository: {
        getters: {
          versionEnabled: jest.fn(() => true),
          authenticated: jest.fn(() => true),
          getUpdateVersionsState: jest.fn(() => true),
          getCurrentRepository: jest.fn(() => currentRepository),
          getNameVersion: jest.fn(() => 'bothub'),
        },
      },
    },
  });
  let wrapper;
  const $router = {
    currentRoute: {
      name: 'repository-summary',
    },
  };
  beforeEach(() => {
    wrapper = shallowMount(SideBar, {
      localVue,
      store,
      mocks: {
        $t: () => 'some specific text',
        $router
      },
    });
  });

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
