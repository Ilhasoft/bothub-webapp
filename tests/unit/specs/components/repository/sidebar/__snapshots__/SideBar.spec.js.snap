// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar.vue renders correctly 1`] = `undefined`;
